sky/backends/backend_utils.py

```diff
@@ -1822,26 +1837,32 @@ def run_ray_status_to_check_ray_cluster_healthy() -> bool:
                 raise exceptions.FetchClusterInfoError(
                     reason=exceptions.FetchClusterInfoError.Reason.HEAD)
             head_runner = runners[0]
-            rc, output, stderr = head_runner.run(
-                instance_setup.RAY_STATUS_WITH_SKY_RAY_PORT_COMMAND,
-                stream_logs=False,
-                require_outputs=True,
-                separate_stderr=True)
-            if rc:
-                raise RuntimeError(
-                    f'Refreshing status ({cluster_name!r}): Failed to check '
-                    f'ray cluster\'s healthiness with '
-                    f'{instance_setup.RAY_STATUS_WITH_SKY_RAY_PORT_COMMAND}.\n'
-                    f'-- stdout --\n{output}\n-- stderr --\n{stderr}')
 
-            ready_head, ready_workers = _count_healthy_nodes_from_ray(output)
+            ready_head, ready_workers, output, stderr = (
+                get_node_counts_from_ray_status(head_runner))
             total_nodes = handle.launched_nodes * handle.num_ips_per_node
-            if ready_head + ready_workers == total_nodes:
-                return True
-            raise RuntimeError(
-                f'Refreshing status ({cluster_name!r}): ray status not showing '
-                f'all nodes ({ready_head + ready_workers}/'
-                f'{total_nodes}); output: {output}; stderr: {stderr}')
+            if ready_head + ready_workers != total_nodes:
+                # If cluster JUST started, maybe not all the nodes have shown up.
+                # Wait for a few seconds and try again.
+                # Note: We are okay with this performance hit because it's very
+                # rare to normally hit this case. It requires:
+                # - All the instances in the cluster are up on the cloud side
+                #   (not preempted), but
+                # - The ray cluster is somehow degraded so not all instances are
+                #   showing up
+                logger.debug(f'Refreshing status ({cluster_name!r}): first '
+                             'ray status not showing all nodes '
+                             f'({ready_head + ready_workers}/{total_nodes}); '
+                             f'output:\n{output}\nstderr:\n{stderr}')
+                time.sleep(5)
```

Comment:
wondering if a polling based method with a slightly larger timeout would make it more robust and efficient here.

Reply:
Oh, I mean a polling method could return earlier than 5s if the `ray status` becomes valid.

Also, we had a wait for the `ray status` to show something in the ray cluster setup during the instance_setup.py below. I assume this PR resolves the case where ray status takes longer to be valid in multi-node case.
https://github.com/skypilot-org/skypilot/blob/f2c2092f3fc343d59348bf209b3ad0d4d76cea7b/sky/provision/instance_setup.py#L59-L63

Reply:
Actually k8s also seems fine.